Task03

stenkarazin2 opened the pull request.

stenkarazin2 committed:
Task03

stvort commented on z03/src/main/java/ru/otus/springboot/service/AnswersServiceImpl.java:
```diff
@@ -0,0 +1,67 @@
+        List< AnsweredTestItem > answeredTestItemList = new LinkedList<AnsweredTestItem>();
+
+        TestItem testItem;
+        for ( int j = 0; j < testItemListSize; j++ ) {
```
Похоже, что тут может быть for in

stenkarazin2 replied:
Я не нумерую задания, а вообще-то это делать надо. Так что оставлю for со счетчиком

stvort commented on z03/src/main/java/ru/otus/springboot/service/AnswersServiceImpl.java:
```diff
@@ -0,0 +1,67 @@
+            ioService.outputString( "\n----------" );
+            ioService.outputString(testItem.getQuestion());
+            List<Variant> variantList = testItem.getVariantList();
+            int i = 0;
+            for ( Variant variant : variantList ) {
```
А тут как раз for со счетчиком

stenkarazin2 replied:
Исправил

stvort commented on z03/src/main/java/ru/otus/springboot/service/AnswersServiceImpl.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    public List< AnsweredTestItem > inputAnswers( List< TestItem > testItemList ) {
+        Integer testItemListSize = testItemList.size();
```
Вместо оберток лучше использовать примитивы если не подразумевается хранение в коллекциях или значение `null`

stenkarazin2 replied:
Исправил

stvort commented on z03/src/main/java/ru/otus/springboot/service/AnswersServiceImpl.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    public void printResults( List< AnsweredTestItem > answeredTestItemList ) {
+        Integer answeredTestItemListSize = answeredTestItemList.size();
+        Integer rightAnsweredTestItemCount = 0;
```
И т.д.

stvort commented on z03/src/main/java/ru/otus/springboot/service/AnswersServiceImpl.java:
```diff
@@ -0,0 +1,67 @@
+                rightAnsweredTestItemCount++;
+            }
+        }
+        ioService.outputString(  "Результат: Верно " + rightAnsweredTestItemCount + " из " + answeredTestItemListSize );
```
По заданию все сообщения должны быть переведены.

Spring boot автоматически, на основании наших настроек в application.yml (`spring.messages.*`) формирует бин `MessageSource`, который будет работать с бандлом, что мы в этих настройках укажем (файл бандла по умолчанию создавать нужно обязательно + по одному для каждого языка). После этого можно указать в настройках локаль приложения и сделать сервис локализации, который обернет собой бин с настройками локали и собственно `MessageSource`. Кстати, ошибкам связанным с локализацией, мы посвятили некоторое время на разборе ДЗ в первом блоке курса

stenkarazin2 replied:
Сделано

stvort commented on z03/src/main/java/ru/otus/springboot/dao/TestBoxDao.java:
```diff
@@ -0,0 +1,11 @@
+
+public interface TestBoxDao {
+
+    List<TestItem> getTestItemList( /*String filename*/ );
```
Мы договаривались не хранить не актуальный код в репозитории

stenkarazin2 replied:
Стараюсь, не всегда получается

stvort commented on z03/src/main/java/ru/otus/springboot/dao/TestBoxDaoSimple.java:
```diff
@@ -0,0 +1,99 @@
+    private String getTextTranslation( String json, String locale ) {
+        Map< String, String > m = new HashMap< String, String >();
+        // Формирование reg1 и reg2 при каждом запуске метода - не лучшее решение с точки зрения производительности
+        String reg1 = "^" + "\\" + "{" + "?" + "\\" + "s" + "*" + "\"" + "{" + "1" + "}";
+        String reg2 = "\"" + "{" + "1" + "}" + "\\" + "s" + "*" + "\\" + "}" + "?" + "$";
```
Так, при увеличении количества языков, строка будет расти и станет не человекочитаемой. Лучше сделать для каждого языка свою строку в классическом CSV-формате, просто с разделителями

stenkarazin2 replied:
Сделал, как сказали. Замечу, что в таком случае теряется связь между формулировками одного и того же вопроса на разных языках

stvort commented on z03/src/main/java/ru/otus/springboot/dao/TestBoxDaoSimple.java:
```diff
@@ -0,0 +1,99 @@
+                        continue;
+                    }
+                    List< Variant > variantList = new ArrayList<>();
+                    Boolean isRight = false;
```
Обертки)

stvort commented on z03/src/main/java/ru/otus/springboot/dao/TestBoxDaoSimple.java:
```diff
@@ -0,0 +1,99 @@
+                    testItemList.add( testItem );
+                }
+            } catch( IOException e ) {
+                // Можно обернуть исключение и бросить выше
```
Нужно) Пока мы просто подавили исключение, чего делать не рекомендуется

stenkarazin2 replied:
Сделал. Теперь все исключения, приходящие с низших уровней, на уровне сервисов обернуты в единое исключение

stvort commented on z03/src/main/java/ru/otus/springboot/dao/TestBoxDaoSimple.java:
```diff
@@ -0,0 +1,99 @@
+    private final String fileName;
+    private final String locale;
+
+    public TestBoxDaoSimple( PropertyConfig propertyConfig ) {
+        fileName = propertyConfig.getFileName();
+        locale = propertyConfig.getLocale();
+    }
```
Дао не нужно знать про все настройки. Тот же `requiredTestItemListSize` ему видимо не особо нужен, а значит мы потенциально универсальный класс намертво приделали к одному приложению, точнее ко всем, у кого есть `requiredTestItemListSize`

stenkarazin2 replied:
Исправил. Теперь requiredTestItemListSize фигурирует на более высоком уровне - уровне логики

stvort commented on z03/src/main/java/ru/otus/springboot/logic/TestBoxImpl.java:
```diff
@@ -0,0 +1,37 @@
+        final List<TestItem> testItemList = testBoxDao.getTestItemList();
+        final Integer testItemListSize = testItemList.size();
+        if( testItemListSize != requiredTestItemListSize ) {
+            // Какие-то действия
+            // Можно также реализовать с помощью исключений
```
Вместо `Какие-то действия` лучше написать эти действия

stvort commented on z03/src/main/java/ru/otus/springboot/service/AnswersServiceImpl.java:
```diff
@@ -0,0 +1,67 @@
+        }
+        ioService.outputString(  "Результат: Верно " + rightAnsweredTestItemCount + " из " + answeredTestItemListSize );
+
+        // Оставшуюся часть метода - перечисление -  можно реализовать как отдельный метод
```
Опять же. Нужно реализовать

stenkarazin2 replied:
Реализовано

stvort commented on z03/src/main/java/ru/otus/springboot/logic/TestBoxImpl.java:
```diff
@@ -0,0 +1,37 @@
+        }
+
+        Collections.shuffle( testItemList );
+        for( int i = requiredTestItemListSize; i < testItemListSize; i++ ) {
+            testItemList.remove( i );
+        }
```
Это по идее не должно работать

stenkarazin2 replied:
Не понял

stenkarazin2 committed:
Implemented message localization

stenkarazin2 committed:
Refactored data localization

stenkarazin2 committed:
Added the only service-level exception NoCinemaException

stenkarazin2 committed:
Added getProperty( String ) to PropertyConfig

Added method getProperty( String property ) to PropertyConfig interface as unified entry for property file application.yml

stenkarazin2 committed:
Transferred requiredTestItemListSize initialization down to middle (logic) layer

stenkarazin2 committed:
Added IllegalArgumentException to method getProperty( String )

stenkarazin2 committed:
Transformed LocaleHolder and renamed variables locale to localeName

stenkarazin2 committed:
Added class NoCinema

stenkarazin2 committed:
Added 2 tests

stenkarazin2 committed:
Moved variables initialization from construcnors

stenkarazin2 committed:
Removed tests

stvort commented on z03/src/main/java/ru/otus/springboot/config/PropertyConfigImpl.java:
```diff
@@ -0,0 +1,47 @@
+    public String getProperty( String property ) throws IllegalArgumentException {
+        // Пользователю класса PropertyConfig предоставляется данный унифицированный метод для обращения к файлу свойств application.yml
+        // Достоинство выбранного подхода состоит в том, что пользователь класса PropertyConfig обращается к свойству по имени, передавая это имя данному унифицированному методу
+        // С другой стороны, на пользователя класса PropertyConfig перекладывается задача интерпретации свойства, т.е. приведения полученного значения к нужному типу данных
```
Он уже есть. Называется `Environment`. И не пользуются им т.к. это не удобно, т.к. 
> на пользователя перекладывается задача интерпретации свойства, т.е. приведения полученного значения к нужному типу данных

stvort commented on z03/src/main/java/ru/otus/springboot/dao/TestBoxDaoSimple.java:
```diff
@@ -0,0 +1,84 @@
+
+    public List< TestItem > getTestItemList() throws IOException, NullPointerException, IllegalArgumentException {
+        String fileName = propertyConfig.getProperty("config-file");
+        String localeName = localeHolder.getLocaleName();
```
Если использовать дао в другом приложении, то там должен быть реализован `PropertyConfig`, чего странно от него требовать, т.к. там может быть свой подход к пропертям. А вот если сделать мелкий интерфейс `FileNameProvider`, то его скорее всего легко реализуют.

stenkarazin2 replied:
Сделано

stvort commented on z03/src/main/java/ru/otus/springboot/config/LocaleHolder.java:
```diff
@@ -0,0 +1,36 @@
+import java.util.Locale;
+
+@Component
+public class LocaleHolder {
```
Поэтому, эту штуку тоже лучше передавать по интерфейсу

stenkarazin2 replied:
Сделано

stvort commented on z03/src/main/java/ru/otus/springboot/dao/TestBoxDaoSimple.java:
```diff
@@ -0,0 +1,84 @@
+import java.util.List;
+
+@Component
+public class TestBoxDaoSimple implements TestBoxDao {
```
Это скорее все же `TestBoxDaoLocalizedCsv`

stvort commented on z03/src/main/java/ru/otus/springboot/dao/TestBoxDao.java:
```diff
@@ -0,0 +1,12 @@
+
+public interface TestBoxDao {
+
+    List<TestItem> getTestItemList() throws IOException, NullPointerException, IllegalArgumentException;
```
Мы многократно на занятиях обсуждали, что дао не может кидать такого рода исключения исходя из самой сути этого паттерна и принципа инверсии зависимостей

stvort commented on z03/src/main/java/ru/otus/springboot/dao/TestBoxDaoSimple.java:
```diff
@@ -0,0 +1,84 @@
+        this.localeHolder = localeHolder;
+    }
+
+    public List< TestItem > getTestItemList() throws IOException, NullPointerException, IllegalArgumentException {
```
Очень большой и сложный для восприятия метод. Его лучше разбить на приватные

stenkarazin2 replied:
Сделано

stvort commented on z03/src/main/java/ru/otus/springboot/logic/TestBoxImpl.java:
```diff
@@ -0,0 +1,61 @@
+        Collections.shuffle( testItemList );
+        for( int i = requiredTestItemListSize; i < testItemListSize; i++ ) {
+            testItemList.remove( i );
+        }
```
>Это по идее не должно работать
>Не понял

Там даже идея подсказывает. Вот пример. Попробуйте выполнить:
```
		var lst = new ArrayList<>(List.of(1, 2, 3, 4, 5, 6, 7, 8));
		var listSize = lst.size();
		for( int i = 5; i < listSize; i++ ) {
			lst.remove( i );
		}
```
Получим
> Exception in thread "main" java.lang.IndexOutOfBoundsException: Index 7 out of bounds for length 6

stvort commented on z03/src/main/java/ru/otus/springboot/service/NoCinema.java:
```diff
@@ -0,0 +1,20 @@
+
+import org.springframework.stereotype.Service;
+
+@Service
+public class NoCinema {
+    private final IOService ioService;
```
Зачем целый сервис, только для того, чтобы напечатать одно сообщение?

stenkarazin2 replied:
Чтобы соответствовать принципу единственной ответственности

stvort commented on z03/src/main/java/ru/otus/springboot/service/MessageSourceWrapper.java:
```diff
@@ -0,0 +1,42 @@
+
+@Service
+public class MessageSourceWrapper {
+    private final ResourceBundleMessageSource messageSource;
```
Внедряем и используем по интерфейсу

stvort commented on z03/src/main/java/ru/otus/springboot/service/MessageSourceWrapper.java:
```diff
@@ -0,0 +1,42 @@
+    private final LocaleHolder localeHolder;
+
+    public MessageSourceWrapper( LocaleHolder localeHolder ) {
+        this.messageSource = new ResourceBundleMessageSource();
+        this.messageSource.setBasenames( "messages" );
```
По заданию `MessageSource` должен создать за нас SpringBoot, мы его только настраиваем через application.yml и внедряем

stenkarazin2 replied:
Сделано

stvort commented on z03/src/main/java/ru/otus/springboot/service/MessageSourceWrapper.java:
```diff
@@ -0,0 +1,42 @@
+        }
+    }
+
+    public String getEmergencyMessage( String messageCode ) {
+        // Причиной исключения может быть локаль
+        // Поэтому нужно разомкнуть круг
+        try {
+            return messageSource.getMessage( messageCode, null, localeHolder.getLocale() );
+        }
+        catch( IllegalArgumentException e ) {
+            return messageSource.getMessage( messageCode, null, Locale.getDefault() );
+        }
+    }
```
Для этого не нужен отдельный метод

stenkarazin2 replied:
Исправлено

stenkarazin2 committed:
Turned on autoconfiguration for MesageSource

stenkarazin2 committed:
Renamed TestBoxDaoSimple to TestBoxDaoLocalizedCsv

stenkarazin2 committed:
Transformed LocaleHolder to interface

stenkarazin2 committed:
Get rid of the bean PropertyConfig in LocaleHolderImpl, TestBoxDaoLocalizedCsv

stenkarazin2 committed:
Transferred NoCinemaException from middle (logic) layer down to dao layer

stenkarazin2 committed:
Improved deletion

stenkarazin2 committed:
Introduced private methods to TestBoxDaoLocalizedCsv

stenkarazin2 committed:
Get rid of the bean PropertyConfig in TestBoxImpl and removed one

stenkarazin2 committed:
Removed method getEmergencyMessage( String )

stenkarazin2 committed:
Added tests

stenkarazin2 committed:
Removed constructor public attribute

stenkarazin2 committed:
Renamed ru.otus.springboot to ru.otus.spring

stvort commented on z03/src/main/java/ru/otus/spring/domain/Variant.java:
```diff
@@ -0,0 +1,22 @@
+public class Variant {
+
+    private final String variantText;
+    private final Boolean isRight;
```
>Вместо оберток лучше использовать примитивы если не подразумевается хранение в коллекциях или значение null

Не исправлено

stvort commented on z03/src/main/java/ru/otus/spring/domain/AnsweredTestItem.java:
```diff
@@ -0,0 +1,27 @@
+public class AnsweredTestItem {
+
+    private final TestItem testItem;
+    private final Integer userAnswer;       // Номер варианта ответа, введенный пользователем
```
И т.д.

stvort commented on z03/src/main/java/ru/otus/spring/config/RequiredTestItemListSizeProvider.java:
```diff
@@ -0,0 +1,7 @@
+
+public interface RequiredTestItemListSizeProvider {
+
+    int getRequiredTestItemListSize() throws NumberFormatException;
```
Точно не возможно ни одной реализации, которая бы не кидала `NumberFormatException`? Скорее тут все же нарушен DIP

stvort commented on z03/src/main/java/ru/otus/spring/dao/TestBoxDaoLocalizedCsv.java:
```diff
@@ -0,0 +1,102 @@
+        return new Variant( str, isRight );
+    }
+
+    public List< TestItem > getTestItemList() throws NoCinemaException {
```
"Кина не будет" достаточно смешное исключение. Но вряд ли юмор уместен в коде. По крайней мере прямой)

stvort commented on z03/src/main/java/ru/otus/spring/dao/TestBoxDaoLocalizedCsv.java:
```diff
@@ -0,0 +1,102 @@
+                // В каждой строке сначала записан код языка (en, ru и т.д.), затем вопрос, а затем варианты ответа, один из которых правильный
+                // Например: ru,Вопрос,Вариант1,,Вариант2,Вариант3
+                // Правильный вариант предваряется двумя запятыми, т.е. в примере выше правильным является Вариант2
+                if( !isRightFormat( line ) ) {
+                    // Bad format
+                    continue;
+                }
+
+                LinkedList< String > snippetList = Arrays.stream( line.split( "," ) )
+                        .collect( Collectors.toCollection( LinkedList::new ) );
+                if( !isRightFormat2( snippetList ) ) {
+                    // Bad format
+                    continue;
+                }
+
+                if( !localeName.equals( snippetList.removeFirst() ) ) {
+                    // Another language
+                    continue;
+                }
```
Это же по идее все исключительные ситуации

stenkarazin2 replied:
Если строка в файле относится к другому языку, то нет, не исключительная. А если неверный формат строки, то можно либо кинуть исключение, либо написать об этом в лог

stvort commented on z03/src/main/java/ru/otus/spring/service/AnswersService.java:
```diff
@@ -0,0 +1,15 @@
+
+public interface AnswersService {
+
+    List< AnsweredTestItem > inputAnswers( List< TestItem > testItemList ) throws NoCinemaException;
+    void printResults( List< AnsweredTestItem > answeredTestItemList ) throws NoCinemaException;
+    void printResultsWidely( List< AnsweredTestItem > answeredTestItemList ) throws NoCinemaException;
```
Ну и одно исключение на все приложение, это тоже самое, что просто написать `Exception`. Исключения, своим типом, должны детализировать ошибку. Высокоуровневые, детализируют с т.з. предметной области

stvort commented on z03/src/main/java/ru/otus/spring/service/NoCinema.java:
```diff
@@ -0,0 +1,20 @@
+import org.springframework.stereotype.Service;
+
+@Service
+public class NoCinema {
```
А значит, если сделать свою систему бизнес-исключений, такой класс просто будет не нужен

stvort commented on z03/src/main/java/ru/otus/spring/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,29 @@
+    public String inputUserName() throws NoCinemaException {
+        String userName = ioService.readStringWithPrompt( messages.getMessage( "input-username" ) );
+        // Яйцо пасхальное обыкновенное :-)
+        userName = ( userName.equals( "" ) ? "Христо Стоичков" : userName );
```
"Христо Стоичков" лучше сделать настройкой

stenkarazin2 replied:
Так просто было удобнее тестировать вручную. Оставил ради шутки, вообще-то надо было убрать

stvort commented on z03/src/test/java/ru/otus/spring/service/UserServiceImplTest.java:
```diff
@@ -0,0 +1,52 @@
+
+        UserService userService = new UserServiceImpl( mockedIOService, mockedMessageSourceWrapper );
+
+        try {
+            String str = userService.inputUserName();
+        }
+        catch ( NoCinemaException e ) {
+        }
```
Зачем это? Подавлять исключения почти никогда не нужно, а в тестах это нужно еще реже. Мы либо ожидаем исключение и проверяем, что оно нужных характеристик, либо это неожиданное исключение и тест должен упасть

stvort commented on z03/src/test/java/ru/otus/spring/service/UserServiceImplTest.java:
```diff
@@ -0,0 +1,52 @@
+        verify( mockedMessageSourceWrapper, times( 1 ) ).getMessage( any() );
+        verify( mockedIOService, times( 1 ) ).readStringWithPrompt( any() );
+
+        try {
+            assertEquals( userName, userService.inputUserName() );
+        }
+        catch ( NoCinemaException e ) {
+        }
```
И т.д.
